### ISSUE TITLE
Constraint-set-aware bidirectional-inference

### ISSUE THREAD

**Issue by ibraheemdev** [astral-sh/ty#3933]
Bidirectional inference works very well when we have concrete type context to provide during expression inference. However, during generic call inference, we are forced to eagerly specialize parameter types in order to provide useful type context. This can lead to us conflating the upper bound on a type variable, based on the annotated type of the call expression, with the upper bound on an argument, which is not necessarily a transitive relationship, due to variance. After https://github.com/astral-sh/ruff/pull/26141, we now also consider constraints introduced by sibling arguments, which makes the problem worse.

Consider the following example:
```py
class Parent: ...
class Child(Parent): ...

def head[T](x: list[T]) -> T:
    return x[0]

x: Parent = head(reveal_type([Child()])) # revealed: list[Parent]
```

We use the outer constraint `T <: Parent` to eagerly specialize the type context of the list literal to `list[Parent]`. While this does not lead to any type errors, it means that we unnecessarily widen the type of the list, which is invariant over `T`, based on its upper bound. In other examples, this may actually lead to false positives:

```py
def f[T](x: T, y: list[T], z: list[T]) -> Sequence[T]:
    return [x]

def _(x: int, z: list[int]) -> Sequence[int | str]:
    return f(x, [x], z) # error[invalid-argument-type]: Expected `list[int | str]`, found `list[int]`
```

Notice that the constraint `T <: int | str` does not imply `list[T] <: list[int | str]`; `list` is invariant, and `T` may materialize to any subtype of `int | str`, not necessarily its upper bound. However, in the case that we instead had a lower bound of `T >: int | str`, the type context of `list[int | str]` allows us to infer the correct types:

```py
class Contra[T]:
    def pop(self) -> T:
        raise NotImplementedError

def f[T](x: list[T]) -> Contra[T]:
    return Contra()

x: Contra[int | str] = f([1])
```

Similarly, in cases where the type variable does materialize to its upper bound, ignoring it entirely may lead to losing useful type context, such as in the case of typed dicts, which have a very narrow subtyping range, often necessitating the inference of a typed dict:

```py
from typing import Sequence, TypedDict

class TD(TypedDict):
    x: int

def f[T](x: list[T]) -> list[T]:
    return x

x: Sequence[TD | None] = f([{"x": 1}])
reveal_type(x)  # revealed: list[TD | None]`
```

In this example, we do not want to widen unnecessarily to `list[TD | None]`, but also do not want to lose the type context of `TD | None` for the dictionary literals.

Eager specialization currently conflates all of these cases, as there is a disconnect between the inference of the outer generic call and its arguments. Closing that gap to correctly unify the constraints requires either propagating constraints down from the outer call, or propagating constraints up from the inner one:
1. During generic call inference, infer constraint-sets for these expressions when we detect that they are being inferred against unspecialized type variables, whose constraints are yet unknown. For example, we would infer the list literal in the final example to be `∃T, list[T] where T :> DictLiteral[{"x": 1}]`, allowing the list to eventually materialize to a `list[TD]`, but also to `list[TD | None]` if it was annotated as such, or any other compatible type, as necessary.
2. Avoid eagerly specializing parameter types, and instead propagate constraints during bidirectional inference. The list literal would then receive the type context `list[T] where T <: TD | None`, allowing it to propagate the type context of `T <: TD | None` to the dictionary literal elements, while not eagerly widening the type of the list literal from `list[TD]`

Note that it is not exactly clear how we would represent the type of the dictionary literal in the first approach. In fact, we would need to be able to infer constraint-sets for arbitrary expressions and operations, which becomes trick, e.g., during lambda inference. Additionally, we would likely need to re-infer all types after generic call inference, in order to emit diagnostics that do not exist until the unified constraint-set is solved.

The second approach, propagating constraints during bidirectional-inference, is closer to what we do currently, and seems more straightforward to implement. However, the constraints we infer during earlier rounds are not necessarily correct, and we suffer from pollution of argument constraints during fixpoint iteration (see https://github.com/astral-sh/ty/issues/3934), a problem that inferring constraint-sets would avoid. As noted on https://github.com/astral-sh/ty/issues/3934, that problem might also be solved by a more effective propagating of constraints during bidirectional-inference, without requiring the full constraint-set inference except in specific cases (e.g., typed dicts), but there is definitely an overlap between both approaches.

**Comment by jenshnielsen** [comment]
I have run into another inference issue with defaults for type parameters. I am not sure if it is covered by this issue but figured I would add it here before making a new issue. Please let me know if this is better as a new issue.

```python
class Box[T]:
    pass

def make[T = int](cls: type[T] | None = None) -> Box[T]:
    raise NotImplementedError

def caller() -> None:
    a: Box[str] = make()
```

Results in (ty 0.73 python 3.13 windows 11)

```
error[invalid-assignment]: Object of type `Box[int]` is not assignable to `Box[str]`
 --> repro.py:8:19
  |
8 |     a: Box[str] = make()
  |                   ^^^^^^
```

Mypy and pyright both type checks this code without any issues.

**Comment by carljm** [comment]
@jenshnielsen I think that actually is closer to #3559 and will be fixed by astral-sh/ruff#26680.
